pkg/logging/logging.go

```diff
@@ -14,29 +14,77 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Package logging provides a logger that satisfies https://github.com/go-logr/logr.
-// It is implemented as a light wrapper around sigs.k8s.io/controller-runtime/pkg/log
+// Package logging provides Crossplane's recommended logging interface.
+//
+// The logging interface defined by this package is inspired by Dave Cheney's
+// "Let’s talk about logging" article:
+//
+// https://dave.cheney.net/2015/11/05/lets-talk-about-logging
+//
+// It is similar to other logging interfaces inspired by said article, namely:
+//
+// https://github.com/go-logr/logr
+// https://github.com/go-log/log
+//
+// Crossplane prefers not to use go-logr because it desires a simpler API with
+// only two levels (per Dave's article); Info and Debug. Crossplane prefers not
+// to use go-log because it does not support structured logging. This package
+// *is* however a subset of go-logr's functionality, and is intended to wrap
+// go-logr (interfaces all the way down!), in order to maintain compatibility
+// with the https://github.com/kubernetes-sigs/controller-runtime/ log plumbing.
 package logging
 
-import (
-	"github.com/go-logr/logr"
-	runtimelog "sigs.k8s.io/controller-runtime/pkg/log"
-)
+import "github.com/go-logr/logr"
 
-// Logging levels.
-const (
-	Debug = 1
-)
+// A Logger logs messages. Messages may be supplemented by structured data.
+type Logger interface {
+	// Info logs a message with optional structured data. Structured data must
+	// be supplied as an array that alternates between string keys and values of
+	// an arbitrary type. Use Info for messages that Crossplane operators are
+	// very likely to be concerned with when running Crossplane.
+	Info(msg string, keysAndValues ...interface{})
 
-var (
-	logger = runtimelog.Log
+	// Debug logs a message with optional structured data. Structured data must
+	// be supplied as an array that alternates between string keys and values of
+	// an arbitrary type. Use Debug for messages that Crossplane operators or
+	// developers may be concerned with when debugging Crossplane.
+	Debug(msg string, keysAndValues ...interface{})
 
-	// Logger is the base logger used by Crossplane. It delegates to another
-	// logr.Logger. You *must* call SetLogger to get any actual logging.
-	Logger = logger.WithName("crossplane")
-)
+	// WithValues returns a Logger that will include the supplied structured
+	// data with any subsequent messages it logs. Structured data must
+	// be supplied as an array that alternates between string keys and values of
+	// an arbitrary type.
+	WithValues(keysAndValues ...interface{}) Logger
+}
+
+// NewNopLogger returns a Logger that does nothing.
+func NewNopLogger() Logger { return nopLogger{} }
```

Comment:
I'm really pulling for the version of the future where we use "Nope Logger" instead of "Nop Logger"

Reply:
https://godoc.org/go.uber.org/zap#NewNop

FWIW `NopFoo` is a fairly typical name for no-op implementations of instrumentation in Go.